Deadlines

frisi opened the pull request:
hi guido

i started working on collective.contentrule.mustread (which uses the schedule_must_read api of collective.mustread). i'll add a github repo for this soon so you can see how the api is used there.

in the meantime i'd love to get your feedback on this pr.
especially discuss the comments marked with `@guido` (as `TODO` did not survive code-analysis ;-)

frisi committed:
add setup profile and step to create database file and configure it in the registry

frisi committed:
implement api for schedulung mustread requests

* refine interface definition
* add new columns to mustread db table (and provide upgrade step that adds them)
* add unit tests for new functionality

frisi commented:
@gyst tests pass, but the travis build seems to not terminate properly. could you please assist here, i don't know the ci setup well enough

ale-rt commented:
Related #4

ale-rt commented:
I rebased it on master after merging #4.

frisi committed:
fix csv report

gyst commented on README.rst:
```diff
@@ -70,6 +70,21 @@ The database storage provides a rich API as specified in ``collective.mustread.i
 
 Note that the API design already contains specifications for some future extensions that have not been implemented yet.
 
```
Is that still true?

frisi replied:
good catch - i'll rip that off

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -59,22 +59,22 @@ class ITracker(Interface):
-    not even possible) to schedule a ``must_read(obj)`` to track reads.
+    - uids_read('johndoe')
 
     Scheduled for later extension:
```
I guess this is now in and the heading should be removed?

frisi replied:
guess you meant `Scheduled for later extension:` . will remove that

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
 
+        @guido: i'd suggest to remove force_deadline as we have the get_stats
+        method to get a detailed report which users read and when
+
```
This was added as per your spec. I'm neutral, but if you want to remove `force_deadline` here IMO the `deadline` arg should be removed from `schedule_must_read` also. Either this application has a deadline concept and a way to query for that, or else the whole deadline thingy is handled by consuming code based on datetime reporting and this application itself does not care.

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
+
+        @guido: if we implement the force_deadline here, we'd have an easy way
+        to list all items with missed deadlines too. alternatively we can make
+        get_report more powerful/complex
```
If `deadline` is a meaningful concept in the business domain, as it appears to be, I'm in favor of adding it here as well. I think it's great to have well supported code paths to enforce read-before-deadline without having to write report parsers in consuming code.

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
+        :returns: List of content objects with open read requests
+        :rtype: Generator
+        '''
+        # xxx guido: generator or list? (we intend to use all elements anyway)
```
Even if you use them all, if using means applying a filter and discarding 90% of entries, then a generator is more efficient.

frisi replied:
taken... as what_to_read returns only a few objects (compared to the report) i'll make it return a list and stay with the generator for the report

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
+        '''Maintenance method to remove open must-read requests for an object
+        and/or a specified userid
+
+        @guido decide if we need/want this
```
LOL you decide! I don't need `schedule_must_read` myself so I also don't need `unschedule_must_read`.

frisi replied:
i can imagine that when using the mustread functionality a lot you'll end up with quite a lot of objects returned by what_to_read although their deadline has passed a long time ago (think of employees that quit, where ill and "forget" about a reminder, etc).
we'll need a view (eg in collective.contentrules.mustread) that lists open requests and also allows privileged users to undo the schedulung of an open request or mark an open request as "accepted"

gyst replied:
For that use case it makes more sense to introduce a `scheduled_after` or `deadline_after` arg to the relevant query methods.

frisi replied:
yes, a deadline_after=datetime would make sense for `what_to_read` indeed

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
+        list or accept that a user (that might be ill) will not be required to
+        read a content item
+
+        @guido decide if we need/want this
```
Definitely not, since it duplicates `unschedule_must_read` and the approach suggested in this docstring is far more convoluted.

frisi replied:
@gyst there is a semantic difference between: delete a read request from the database (unschedule_must_read - i.e. because it was scheduled by a mistake) and accepting that a user has not confirmed an open request (accept_not_read) 

however, i also agree, that adding another possible state `accepted` (or whatever we name it) adds extra complexity and is not required for my usecase either. so i'll drop that

gyst commented on src/collective/mustread/models.py:
```diff
@@ -17,10 +17,12 @@ class MustRead(Base):
     read_at = Column(DateTime, nullable=True)
     deadline = Column(DateTime, nullable=True)
+    scheduled_at = Column(DateTime, nullable=True)
+    scheduled_by = Column(String(255))
```
This is very implicit, since it's not part of the interface contract of `schedule_must_read`. Which means you're implicitly adding the `request` to the interface contract of a backend component which is request-agnostic. In other words this column implies you're turning a database utility into a browser view. This will make testing and debugging much more difficult. If you need `scheduled_by`, add it to the contract of `schedule_must_read`.

frisi replied:
good point, i'll add scheduled_by to the signature and will not implicitly use current_user_id

gyst commented on src/collective/mustread/models.py:
```diff
@@ -17,10 +17,12 @@ class MustRead(Base):
-    site_name = Column(Unicode(255), nullable=True)
-    info = Column(Unicode(255), nullable=True)
+    site_name = Column(Unicode(255), nullable=True)  # xxx remove?
+    info = Column(Unicode(255), nullable=True)  # xxx remove?
```
Yes all of that cruft is copied over from the auditlog. Feel free to strip it down.

frisi replied:
done.
title was selected in a query, but did not get accessed in the result set

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
-        :returns: Userids of users scheduled for reading which did not read
-        :rtype: List
+        :returns: Dictionary with userids of users scheduled for reading which
+                  did not read and their deadline (``{userid: deadline}``)
```
Here again you're working with the `deadline` concept, so I say keep `force_deadline` in the method signature for this method.

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -190,10 +192,91 @@ def who_unread(obj, force_deadline=True):
+        If context is None, defaults to plone site.
+
+        If include_children is ``True`` (which is the default), include
+        mustread records for child-objects of context.
```
Huh? The SQL layer has no containment concept. It does not know about Plone objects - those are squashed into UID strings and that's all the SQL ever needs to know about the content. Introducing this will have massive implementation implications, since you now cannot run the report in plain SQL anymore but will have to recurse into the Plone DB for any UID to determine what the child UIDs are, and then recurse again.

gyst replied:
Update: I now see the SQL layer is path aware and you're using that. Maybe that can be made more explicit in this interface contract by saying "child objects whose `path` startswith the context `path`"  or something like that. But I'm OK with this and withdraw my objection above. Just keeping it for reference.

frisi replied:
added `(path starts with context's path)` to the docstring

gyst commented on src/collective/mustread/tests/test_tracker.py:
```diff
@@ -195,12 +195,200 @@ def test_most_read_dates_verylongago_limit(self):
+        """existing mustread request for an object do not get
+        overwritten, we return ids of users a request has
+        been written for
+        """
```
Smart, I overlooked that in the interface docstring but it makes sense.

gyst commented on src/collective/mustread/tracker.py:
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
+                log.info('user {0} already has read / has to read {1}'.format(
+                    userid, path))
+                continue
+            # @guido: should we change the deadline for existing must_reads?
```
Yes that makes sense. However then you should also update the `scheduled_at` (and possibly the `scheduled_by` but see above) and now you're painting yourself into a corner where it's possible that a document was read by a user before it was scheduled (because she read it on the first scheduling, before it got scheduled again). Of course that is already possible anyway: a user may have read an item before it got scheduled at all. Do you have test coverage for that edge case?

frisi replied:
no. and i don't think it's necessary:
it could be that the document has been changed in the meantime and requires to be read again.
i'd say it's okay to require users to read an object after a read request has been sheduled, no matter if they marked it as read before.

gyst replied:
OK but then you'll have to also update the interface contract. Also it then does not make sense to return a user list anymore, since now this will axiomatically apply to all the userids you passed in.

gyst commented on src/collective/mustread/tracker.py:
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
+        uid = utils.getUID(obj)
+        hostname = utils.hostname()
+        now = datetime.utcnow()
+        current_user = api.user.get_current().id
```
This will bite you once the client asks you to do a cron job 'on behalf of' the CEO. See my comments above. But maybe it's not that big a deal.

frisi replied:
you are right, i changed scheduled_by to be explicitly passed as argument to schedule_must_read

gyst commented on src/collective/mustread/tracker.py:
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
+        if userid:
+            query = query.filter(MustRead.userid == userid)
+        if include_children:
+            query = query.filter(MustRead.path.startswith(path))
```
Ah!

gyst commented on src/collective/mustread/tracker.py:
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
+        csvfile is a file-like object with a write method
+        (see csv.writer documentation)
+        xxx if guido does not like csv creation here in the api we can move
+        this to our contentrule
```
Oh no this is cool. We're managing a SQL DB and this is a very valid view on that DB.

frisi replied:
fine, i'll add it to the official api

gyst reviewed (changes requested):
I haven't actually tested this, just reviewed the code, but looks fine to me in general, see more specific comments I made.

The main thing is to NOT force consumers to rely on the report generation for application logic. In other words: you use the report if you want to render a report, but you use the core API if you actually want to query the DB to support a specific execution path. IMO that implies supporting the `deadline` concept systematically throughout the CRUD API.

frisi commented on src/collective/mustread/tracker.py:
```diff
@@ -91,11 +103,117 @@ def most_read(self, days=None, limit=None):
+        csvfile is a file-like object with a write method
+        (see csv.writer documentation)
+        xxx if guido does not like csv creation here in the api we can move
+        this to our contentrule
```
thanks for your detailed feedback @gyst.
i'll iterate over your comments and will call you afterwards if any points are unclear to me

frisi commented:
thanks for your detailed feedback @gyst.
i'll iterate over your comments and will call you afterwards if any points are unclear to me

frisi committed:
improvements after first review of guido

* cleanup documentation
* make what_to_read return a list instead of a generator
* remove 'accept_not_read' from the interface
* add get_report_csv to the official api and complete docstring
* add scheduled_by to schedule_must_read signature (explicit is better than implicitly reading userid from the request)

frisi committed:
fix warning in tests

comparing unicode value with non-unicode would result in warnings such as

  SAWarning: Unicode type received non-unicode bind param value

frisi committed:
remove unnecessary columns

title, site_name and info are not needed for our usecase

frisi commented:
i pushed some commits dealing with your review.
topics still open currently are:

* force_deadline
* whether to update an existing read request when scheduling another one or not.

i'll have a look at this after/while finishing https://github.com/collective/collective.contentrules.mustread

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -152,36 +150,40 @@ def most_read(days=None, limit=None):
         :type obj: Content object (must be IUUID resolvable)
         :param userids: Userids of the users that should view the object.
-        :type userid: List
+        :type userids: List
```
This should be made optional, and then either default to the set of all users at schedule time, or `None` implying all users in the system at report time. Using schedule time is more robust - you don't want a new employee coming in and suddenly having a whole backlog of must_read items.

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -152,36 +150,40 @@ def most_read(days=None, limit=None):
+        :param by: Userid that scheduled the must read request
+        :type by: string
+        :returns: List of userids a mustread deadline has been scheduled for
+                  (existing deadlines do not get changed)
```
If you do decide to re-schedule this return value loses its value and should be dropped.

frisi replied:
i guess it's better to not allow re-scheduling of existing read requests.
instead the frontend needs to delete existing mustreads and create new ones.

gyst reviewed (commented):
See my newer comments on the fallout of updating existing scheduled must reads.

gyst commented on src/collective/mustread/interfaces.py:
```diff
@@ -59,22 +59,20 @@ class ITracker(Interface):
-    - who_unread(obj)
+    - who_must_read(obj)
+    - what_to_read(context)
+    - what_to_read(userid)
```
Please also add `unschedule_must_read` to this docstring.

frisi replied:
done in https://github.com/collective/collective.mustread/pull/1/commits/73f1a3dd808f4ea49ec70f8cce08319cd30e39a2

gyst reviewed (commented):
Yes I think that is the most elegant and a good reason to have `unschedule_must_read`. However this implies that re-scheduling a must-read on a page that was already scheduled should throw an exception, and this needs to be documented in the interface contract and tested for.

But even when not rescheduling the return value should be removed, since again that is axiomatically identical to the method call argument.

frisi committed:
complete high level description

add missing methods to the high level description.

frisi commented:
> Yes I think that is the most elegant and a good reason to have unschedule_must_read. However this implies that re-scheduling a must-read on a page that was already scheduled should throw an exception, and this needs to be documented in the interface contract and tested for.

i'd say we can keep it the way it is:

* an existing read request on page1 for user 'guido'
* call schedule_must_read(page1, ['guido', 'harald'], deadline)
* returns ['harald'], creates a deadline for harald (this must not necessarily be the same as the one of guido) and leaves the record for guido as it is.

this would allow a usecase where you grant additional users a role (or add them to a group) that matches the role-filter for a collective.contentrules.mustread action and re-fire the "request read confirmation event"
new users will get a mustread-db-record and receive a notification email. exisiting records for users will be unharmed.

> But even when not rescheduling the return value should be removed, since again that is axiomatically identical to the method call argument.

currently this is not the case. as we don't re-schedule existing records but skip them, the return value can be a subset of the userids passed to `schedule_must_read` in case there are existing mustread-records for.

actually i see that i'll need to change the current implementation so that records with state `read` are set to `must_read` and `read_at` gets cleared and scheduled_at, deadline, [and optionally scheduled_by] get set.

then we should be fine, shouldn't we?

gyst commented:
> new users will get a mustread-db-record and receive a notification email. exisiting records for users will be unharmed.

Ack.

> as we don't re-schedule existing records but skip them, the return value can be a subset of the userids passed to schedule_must_read in case there are existing mustread-records for.

Ack.

> i'll need to change the current implementation so that records with state read are set to must_read and read_at gets cleared 

This I don't get. It contradicts the policy you've just outlined in the other statements I quoted. Or am I missing something?

frisi commented:
> This I don't get. It contradicts the policy you've just outlined in the other statements I quoted. Or am I missing something?

currently we don't create a record with state mustread if there is an existing record for this object and user.
https://github.com/collective/collective.mustread/blob/deadlines/src/collective/mustread/tracker.py#L113

i'd change this to only skip the record if there is an existing read request (`scheduled_at != None`  no matter if it's marked as read or still mustread)

if the existing record is just a read-marker (`read_at != None`, `scheduled_at==None` then we create a must-read record for the person.

this way we omit the edge-case that no read-requests get sheduled for persons that read the object by accident (implementations such as yours in quaive mark objects as read unnoticed by the end-user) _before_ the read-request got created.

the reliable contract is:
* if a requests got scheduled before, it won't be changed
* if no request got scheduled before, we'll create one

gyst commented:
I got that scenario, but I think it's confusing. If you schedule something as must-read sometime *after* you publish the page, why would that invalidate existing reads? Keeping the existing read record would be consistent with the scenario where we don't update the record if a user read a page between a first must-read schedule and a second must-read call on the same page.

In all of these scenarios, if the admin really wants to force re-reads they can call unschedule_must_read - oh no that's different again. You'd need a `clear_reads()` for that instead :-(.

OK I don't have an easy consistent solution for all these edge cases. And please don't feel blocked by my comments, I'm not making them forcefully but rather playfully since I enjoy doing it. Consider yourself the package maintainer and do it the way seems right to you...

frisi committed:
fix tests when run at 0:30 CEST

frisi committed:
improve api after review

https://github.com/collective/collective.mustread/pull/1#pullrequestreview-32793724

* rename who_must_read to who_did_not_read to cover and additional usecase ("who has missed to read?") via deadline_before
* clarify what happens to existing read requests and read-records when schedule_must_read multiple times for an object/user
  implement tests for this
* improve documentation for unschedule_must_read (not implemented yet)
* add deadline_before arguemnt to what_to_read too

frisi commented:
thanks for your feedback and letting me decide about the (un)schedule api @gyst 

i tried to my best to provide a consitent and slim api. see the commit message for details: https://github.com/collective/collective.mustread/pull/1/commits/b6ca3bc2927e8e9b3ddf78096f0cca8b1edfafcb

i rebased the branch with master and merged my upgrade step with the one @ale-rt provided in the meantime to make it ready to merge.

could you please give me your feedback on the api changes done in the last commit.
if you're ok with that, please merge or let me know that i can do so.

thanks!

frisi committed:
add foce_deadline to what_to_read

frisi commented:
@gyst, i'd like to deploy this today. do you consider it ready to merge into master or do you have any objections?

frisi committed:
support multiple plonesite on one zope instance

limit to objects within plone site in any case to prevent
returning objects not within this plone site.

frisi commented:
thanks for merging
@gyst: could you please make a release?
or grant me pypi privileges? - user frisi

gyst commented:
@frisi I've made you Owner on pypi
